### PR TITLE
New lints `iter_filter_is_some` and `iter_filter_is_ok`

### LINKED ISSUE #11843: Suggest `flatten` over `filter(Option::is_some)`
### What it does

Checks for usage of  `filter(Option::is_some)` or `filter(Result::is_ok)` on an `Iterator`.

This is more succinctly written with just `flatten()`.

Careful: The suggestion must be `MaybeIncorrect`, as this changes the type of the iterator from `Item = Option<T>` to `Item = T`. 

`good-first-issue` instructions: This lint should be quite easy to add as another `clippy_lints/src/methods/` lint. All of the parts required to check for this should already be there and can be reused.

### Advantage

- Shorter code
- Potentially more readable code in following methods in the iterator chain:
    ```rust
    .filter(Option::is_some).map(|x| /*some code*/ x.unwrap() /*some code*/)
    // becomes
    .flatten().map(|x| /*some code*/ x /*some code*/)
    ```

### Drawbacks

Changes the type of the iterator chain. This will lead to more refactorings required down the chain. But I think in most cases this will benefit the code complexity overall.

### Example

```rust
v.into_iter().filter(Option::is_some);
v.into_iter().filter(Result::is_ok);
```

Could be written as:

```rust
v.into_iter().flatten();
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3af174772dfa82c8e1760f4ec0713b4d)

### PR DESCRIPTION
**Pull request by PartiallyTyped**
Adds a pair of lints that check for cases of an iterator over `Result` and `Option` followed by `filter` without being followed by `map` as that is covered already by a different, specialized lint.

Fixes #11843

PS, I also made some minor documentations fixes in a case where a double tick (`) was included.

---

changelog: New Lint: [`iter_filter_is_some`]
[#12004](https://github.com/rust-lang/rust/pull/12004)
changelog: New Lint: [`iter_filter_is_ok`]
[#12004](https://github.com/rust-lang/rust/pull/12004)

### CODE REVIEW

**Comment by PartiallyTyped** [comment]
PS, sorry for the double PR, I was doing Christmas cleaning and deleted my fork. Didn't know GH wouldn't allow me to reopen the previous one 😓

File: clippy_lints/src/methods/mod.rs [hunk 1]
```diff
@@ -3776,6 +3777,58 @@ declare_clippy_lint! {
...
+    /// ### Example
+    /// ```no_run
+    /// // example code where clippy issues a warning
+    /// vec![Ok::<i32, String>(1)].into_iter().filter(Result::is_ok);
+    ///
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// // example code which does not raise clippy warning
+    /// vec![Ok::<i32, String>(1)].into_iter().flatten();
```
Comment (xFrednet) [inline clippy_lints/src/methods/mod.rs:3824]: `flatten()` was added in `1.29.0`. I think a MSRV check is not that important for such an old version, but still good to have. Could you add one?

File: clippy_lints/src/methods/iter_filter.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    if is_iterator && parent_is_not_map && is_method(cx, filter_arg, sym!(is_some)) {
+        span_lint_and_sugg(
+            cx,
+            ITER_FILTER_IS_SOME,
+            filter_span.with_hi(expr.span.hi()),
+            "`filter` for `is_some` on iterator over `Option`",
+            "consider using `flatten` instead",
+            reindent_multiline(Cow::Borrowed("flatten()"), true, indent_of(cx, filter_span)).into_owned(),
+            Applicability::MaybeIncorrect,
```
Comment (xFrednet) [inline clippy_lints/src/methods/iter_filter.rs:65]: [`MaybeIncorrect`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/enum.Applicability.html#variant.MaybeIncorrect) should generally still produce compilable code afterwards. [`HasPlaceholders`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/enum.Applicability.html#variant.HasPlaceholders) would be a better pick here. (Some additional info, if you're interested: The names are not really good in this regard. There is a discussion to change them, but it's not that important, so nobody is really pushing it to it's conclusion)

```suggestion
            Applicability::HasPlaceholders,
```

**Review by xFrednet — COMMENTED** [review]
Generally looks good to me, a few NITs and then it should be good to go. (I hope my review is not too nit-picky.)

File: clippy_lints/src/methods/mod.rs (hunk 1 above)
Comment (PartiallyTyped) [reply clippy_lints/src/methods/mod.rs:3824]: I added the check right before the dispatch, is that okay?

Comment (xFrednet) [reply clippy_lints/src/methods/mod.rs:3824]: Yes, that is good :+1:

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
Looks good to me, thank you for the new lints :D

@bors r+

**Comment by xFrednet** [comment]
Woops, looks like I invalidated the changelog entries. Let's try it again:

@bors retry

**Comment by PartiallyTyped** [comment]
Woops, seems that I didn't squash and force push 😬😔

**Pull request merged**